src/object.c

```diff
@@ -231,31 +260,44 @@ robj *createStringObjectFromSds(const sds s) {
     return createStringObject(s, sdslen(s));
 }
 
-robj *createStringObjectWithKeyAndExpire(const char *ptr, size_t len, const sds key, long long expire) {
-    /* When to embed? Embed when the sum is up to 64 bytes. There may be better
-     * heuristics, e.g. we can look at the jemalloc sizes (16-byte intervals up
-     * to 128 bytes). */
-    size_t size = sizeof(robj);
-    if (key) {
-        size_t key_len = sdslen(key);
-        size += sdsReqSize(key_len, sdsReqType(key_len)) + 1; /* 1 byte for prefixed sds hdr size */
-    }
-    size += (expire != -1) * sizeof(long long);
-    size += sdsReqSize(len, SDS_TYPE_8);
-    if (size <= 64) {
+static robj *createStringObjectWithKeyAndExpire(const char *ptr, size_t len, const sds key, long long expire) {
+    if (shouldEmbedStringObject(len, key, expire)) {
         return createEmbeddedStringObjectWithKeyAndExpire(ptr, len, key, expire);
     } else {
-        return createObjectWithKeyAndExpire(OBJ_STRING, sdsnewlen(ptr, len), key, expire);
+        return createUnembeddedObjectWithKeyAndExpire(OBJ_STRING, sdsnewlen(ptr, len), key, expire);
     }
 }
 
-sds objectGetKey(const robj *val) {
-    unsigned char *data = (void *)(val + 1);
-    if (val->hasexpire) {
+void *objectGetVal(const robj *o) {
+    if (o->hasembval) {
+        unsigned char *data = objectEmbeddedData(o);
+        if (o->hasexpire) {
+            /* Skip expire field */
+            data += sizeof(long long);
+        }
+        if (o->hasembkey) {
+            /* Skip embedded key */
+            uint8_t hdr_size = *(uint8_t *)data;
+            data += 1 + hdr_size;                /* +1 for header size byte */
+            data += sdslen((const_sds)data) + 1; /* +1 for null terminator */
+        }
+        if (o->encoding == OBJ_ENCODING_EMBSTR) {
+            return data + sdsHdrSize(SDS_TYPE_8);
+        } else {
+            return data;
+        }
```

Comment:
Embedded value is only used for OBJ_ENCODING_EMBSTR, right? The else clause is dead code. Better use an assert, or do you have a plan to use embedded values of other encodings?

```suggestion
        assert(o->encoding == OBJ_ENCODING_EMBSTR);
        return data + sdsHdrSize(SDS_TYPE_8);
```